.github/workflows/qns.yml

```diff
@@ -150,20 +148,30 @@ jobs:
           client: ${{ steps.depair.outputs.client }}
           server: ${{ steps.depair.outputs.server }}
           implementations: ${{ needs.implementations.outputs.implementations }}
+          test: handshake,transfer
 
   report:
-    if: ${{ always() && github.event_name == 'pull_request' }}
     name: Report results
     needs: run-qns
     runs-on: ubuntu-latest
     steps:
+      - name: Download cached main-branch results
+        uses: actions/cache/restore@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
+        with:
+          path: results-main
+          key: qns-${{ runner.os }}-${{ github.sha }}
+          restore-keys: qns-${{ runner.os }}-
+
       - uses: actions/checkout@1d96c772d19495a3b5c517cd2bc0cb401ea0529f # v4.1.3
       - uses: actions/download-artifact@9c19ed7fe5d278cd354c7dfd5d3b88589c7e2395 # v4.1.6
         with:
           pattern: '*results'
           path: results
 
       - run: |
+          ls -l results-main || true
```

Comment:
A suggestion only: this is now long and complex enough that it might benefit from being in a file of its own.

Reply:
Let's push this refactoring to the follow-on PR that actually generates the diff?